File: core-scala/src/test/scala/com/baeldung/scala/EmployeeUnitTest.scala
package com.baeldung.scala

import org.junit.Assert.assertEquals
import org.junit.Test

class EmployeeUnitTest {

  @Test
  def whenEmployeeSalaryIncremented_thenCorrectSalary() = {
    val employee = new Employee("John Doe", 1000)
    employee.incrementSalary()
    assertEquals(1020, employee.salary)
  }

  @Test
  def givenEmployee_whenToStringCalled_thenCorrectStringReturned() = {
    val employee = new Employee("John Doe", 1000)
    assertEquals("Employee(name=John Doe, salary=1000)", employee.toString)
  }

  @Test
  def givenEmployeeWithTrait_whenToStringCalled_thenCorrectStringReturned() = {
    val employee =
      new Employee("John Doe", 1000) with UpperCasePrinter
    assertEquals("EMPLOYEE(NAME=JOHN DOE, SALARY=1000)", employee.toString)
  }

}

File: core-java-modules/core-java-lang-oop-2/src/test/java/com/baeldung/copyconstructor/EmployeeUnitTest.java
package com.baeldung.copyconstructor;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;

import java.util.Date;

import org.junit.Test;

public class EmployeeUnitTest {
    @Test
    public void givenCopyConstructor_whenDeepCopy_thenDistinct() {
        Date d1 = new Date(123);
        Employee e1 = new Employee(1, "Baeldung", d1);
        Employee e2 = new Employee(e1);
        assertEquals(d1, e1.getStartDate());
        assertEquals(d1, e2.getStartDate());

        d1.setTime(456);
        assertEquals(d1, e1.getStartDate());
        assertNotEquals(d1, e2.getStartDate());
    }

    @Test
    public void givenCopyMethod_whenCopy_thenDistinct() {
        Date d1 = new Date(123);
        Employee e1 = new Employee(1, "Baeldung", d1);
        Employee e2 = e1.copy();
        assertEquals(d1, e1.getStartDate());
        assertEquals(d1, e2.getStartDate());

        d1.setTime(456);
        assertEquals(d1, e1.getStartDate());
        assertNotEquals(d1, e2.getStartDate());
    }
}
